web/tests/hash_util.test.js

```diff
@@ -87,6 +87,33 @@ run_test("get_current_nth_hash_section", () => {
     assert.equal(hash_parser.get_current_nth_hash_section(3), "");
 });
 
+run_test("test_is_same_server_message", () => {
+    const dm_message_link = "#narrow/dm/9,15-dm/near/43";
+    assert.equal(hash_parser.is_same_server_message(dm_message_link), true);
+
+    const group_message_link = "#narrow/dm/9,16,15-group/near/68";
+    assert.equal(hash_parser.is_same_server_message(group_message_link), true);
+
+    const stream_message_link = "#narrow/stream/8-design/topic/desktop/near/82";
+    assert.equal(hash_parser.is_same_server_message(stream_message_link), true);
+
+    const stream_link = "#narrow/stream/8-design";
+    assert.equal(hash_parser.is_same_server_message(stream_link), false);
+
+    const topic_link = "#narrow/stream/8-design/topic/desktop";
+    assert.equal(hash_parser.is_same_server_message(topic_link), false);
+
+    const dm_link = "#narrow/dm/15-John";
+    assert.equal(hash_parser.is_same_server_message(dm_link), false);
+
+    const search_link = "#narrow/search/database";
+    assert.equal(hash_parser.is_same_server_message(search_link), false);
+
+    const different_server_message_link =
+        "https://fakechat.zulip.org/#narrow/dm/8,1848,2369-group/near/1717378";
+    assert.equal(hash_parser.is_same_server_message(different_server_message_link), false);
+});
```

Comment:
You should include some tests that are just `#` or other short values -- negative indexing could crash with a bad implementation there.